### ISSUE TITLE
Error with facets or parameterized filters on Boolean values ...

### ISSUE THREAD

**Issue by rayvoelker** [simonw/datasette#1671]
I found a strange behavior, and I'm not sure if it's related to views and boolean values perhaps, or if there's something else weird going on here, but I'll provide an example that may help show what I'm seeing happen.

```bash
#!/bin/bash

echo "\"id\",\"expiration_date\"
0,2018-01-04
1,2019-01-05
2,2020-01-06
3,2021-01-07
4,2022-01-08
5,2023-01-09
6,2024-01-10
7,2025-01-11
8,2026-01-12
9,2027-01-13
" > test.csv
csvs-to-sqlite test.csv test.db
sqlite-utils create-view --replace test.db test_view "select id, expiration_date, case when julianday('NOW') >= julianday(expiration_date) then 1 else 0 end as has_expired FROM test"
```

```bash
datasette test.db
```
![image](https://user-images.githubusercontent.com/9308268/159178745-9c6152f7-eac6-4bf9-bef5-a2d63d3ee13f.png)

![image](https://user-images.githubusercontent.com/9308268/159178824-c8952137-270c-42a4-ad1c-f6ad2c51e499.png)

![image](https://user-images.githubusercontent.com/9308268/159178877-23e00b36-443a-43ef-83e5-e0bdddd3fdcd.png)

![image](https://user-images.githubusercontent.com/9308268/159178918-65922cc7-2514-4735-a72d-4904b99976d4.png)

Thanks again and let me know if you want me to provide anything else!

**Comment by simonw** [comment]
Oh this is fascinating! I replicated the bug (thanks for the steps to reproduce) and it looks like this is down to the following:

<img width="1276" alt="image" src="https://user-images.githubusercontent.com/9599/159370986-1e2fc513-6d6c-4a2f-96dd-dccf5a680fe4.png">

Against views, `where has_expired = 1` returns different results from `where has_expired = '1'`

This doesn't happen against tables because of SQLite's [type affinity](https://www.sqlite.org/datatype3.html#type_affinity) mechanism, which handles the type conversion automatically.

**Comment by simonw** [comment]
I wonder if this will be a problem with generated columns, or with SQLite strict tables?

**Comment by simonw** [comment]
Thinking about options for fixing this...

The following query works fine:
```sql
select * from test_view where cast(has_expired as text) = '1'
```
I don't want to start using this for every query, because one of the goals of Datasette is to help people who are learning SQL:
- #1613

If someone clicks on "View and edit SQL" from a filtered table page I don't want them to have to wonder why that `cast` is there.

But... for querying views, the `cast` turns out to be necessary.

So one fix would be to get the SQL generating logic to use casts like this any time it is operating against a view.

An even better fix would be to detect which columns in a view come from a table and which ones might not, and only use casts for the columns that aren't definitely from a table.

The trick I was exploring here might be able to help with that:
- #1293

**Comment by simonw** [comment]
The alternative to using `cast` here would be for Datasette to convert the `"1"` to a `1` in Python code before passing it as a param.

This feels a bit neater to me, but I still then need to solve the problem of how to identify the "type" of a column that I want to use in a query.

**Comment by simonw** [comment]
No, I think I need to use `cast` - I can't think of any way to ask SQLite "for this query, what types are the columns that will come back from it?"

Even the details from the `explain` trick explored in #1293 don't seem to come back with column type information: https://latest.datasette.io/fixtures?sql=explain+select+pk%2C+text1%2C+text2%2C+[name+with+.+and+spaces]+from+searchable_view+where+%22pk%22+%3D+%3Ap0&p0=1

**Comment by simonw** [comment]
Note that Datasette does already have special logic to convert parameters to integers for numeric comparisons like `>`:

https://github.com/simonw/datasette/blob/c4c9dbd0386e46d2bf199f0ed34e4895c98cb78c/datasette/filters.py#L203-L212

**Comment by simonw** [comment]
Also made me realize that this query:
```sql
select * from sortable where sortable > :p0
```
Only works here thanks to the column affinity thing kicking in too: https://latest.datasette.io/fixtures?sql=select+*+from+sortable+where+sortable+%3E+%3Ap0&p0=70
